posts/14-stable-structures.html.pm

```diff
@@ -0,0 +1,671 @@
+  This article explains the philosophy behind the library and how to use it effectively.
+}
+}
+
+
+◊section{
+◊section-title["design-principles"]{Design principles}
+◊epigraph{
+  ◊blockquote[#:cite "https://youtu.be/rX0ItVEVjHc?t=1464"]{
+    ◊p{Understand the data to understand the problem.}
+    ◊footer{Mike Acton, ◊quoted{Data-Oriented Design and C++}, CppCon 2014}
+  }
+}
+
+◊p{
+  Software designs reflect their creators' values.
+  The following principles shaped the ◊code{stable-structures} library design.
+}
+◊ul[#:class "arrows"]{
+  ◊li{
+    ◊em[#:id "radical-simplicity"]{Radical simplicity.}
+    Programming stable memory is significantly easier than working with conventional file systems.
+    The IC solves many issues with which any good storage must deal: data integrity, partial writes, power outages, and atomicity of multiple writes.
+    Even with all these issues sorted out, complicated designs would be hard to implement, debug, and maintain.
+    Each data structure follows the most straightforward design that solves the problem at hand.
+  }
+  ◊li{
+    ◊em[#:id "backward-compatibility"]{Backward compatibility.}
+    Upgrading the library version must preserve the data.
+    All data structures have a metadata section with the layout version.
+    Newer implementations will respect old layouts and should not require data migration.
+  }
+  ◊li{
+    ◊em{No ◊code{pre_upgrade} hooks.}
+    A bug in the ◊code{pre_upgrade} hook can make your canister ◊a[#:href "/posts/01-effective-rust-canisters.html#upgrade-hook-panics"]{non-upgradable}.
+    The best way to avoid this issue is not to have a ◊code{pre_upgrade} hook.
+  }
+  ◊li{
+    ◊em{Limited blast radius.}
+    If a single data structure has a bug, it should not corrupt the contents of other data structures.
+  }
+  ◊li{
+    ◊em{No reallocation.}
+    Moving large amounts of data is expensive and can lead to prohibitively high cycle consumption.
+    All data structures must manage their memory without costly moves.
+  }
+  ◊li{
+    ◊em{Compatibility with ◊a[#:href "https://github.com/WebAssembly/multi-memory/blob/master/proposals/multi-memory/Overview.md"]{multi-memory} WebAssembly.}
+    The design should work when canisters have multiple stable memories since this feature is on the ◊a[#:href "https://forum.dfinity.org/t/proposal-wasm-native-stable-memory/15966#proposal-7"]{IC roadmap}.
+  }
+}
+◊p{
+  As a result of these goals, using the library requires planning and understanding your data shape.
+  For example, you might need to set an ◊a[#:href "#max-size-attribute"]{upper bound} on a value size that you cannot adjust without data migration.
+  Or you might have to decide how many memory pages you want to allocate for canister configuration.
+  There are other libraries with similar goals whose authors chose a slightly different set of traideoffs, such as ◊code-ref["https://crates.io/crates/ic-stable-memory"]{ic-stable-memory}.
```

Comment:
```suggestion
  There are other libraries with similar goals whose authors chose a slightly different set of trade-offs, such as ◊code-ref["https://crates.io/crates/ic-stable-memory"]{ic-stable-memory}.
```

Reply:
My spell checker was happy with both variants, but it seems that trade-off is more common. Fixed.